src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -56,16 +54,34 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
 
         private static readonly object s_lock = new object();
 
-        // In debug builds, force there to be 2 engines. In release builds, use half the number of processors when
-        // there are at least 6. The lower bound is to avoid using multiple engines on systems which aren't servers.
-#pragma warning disable CA1802 // const works for debug, but needs to be static readonly for release
-        private static readonly int s_engineCount =
-#if DEBUG
-            2;
-#else
-            Environment.ProcessorCount >= 6 ? Environment.ProcessorCount / 2 : 1;
-#endif
-#pragma warning restore CA1802
+        private static readonly int s_engineCount = GetEnginesCount();
+
+        private static int GetEnginesCount()
+        {
+            if (uint.TryParse(Environment.GetEnvironmentVariable("DOTNET_SYSTEM_NET_SOCKETS_THREAD_COUNT"), out uint count))
+            {
+                return (int)count;
+            }
+
+            // the data that stands behind this heuristic can be found at https://github.com/dotnet/runtime/pull/35800#issuecomment-624719500
+            // the goal is to have a single epoll thread per every 28 cores
+            const int coresPerSingleEpollThread = 28;
```

Comment:
The benchmarks confirm an observation we had made from the perftraces. It may interesting to put in the comment:

TechEmpower JSON platform benchmark (which has a low workload per request) shows the epoll thread is fully loaded on a 28-core machine. We add 1 epoll thread per 28 cores to avoid it being a bottleneck.

Reply:
I can already hear all the complaints about this line ... me being the first concerned. I think we should not have any heuristic with such a value. What if tomorrow we decide to use a different hardware to do benchmarks? We should have some heuristics that are good for the general cases, and allow customer to define custom values that might be better for them. In our case, in the TE repository we would then define an ENV with a number of epoll thread we want. Same for ARM probably, which might depend on each vendor.

Reply:
> I can already hear all the complaints about this line ... me being the first concerned.

I guess you refer to the suggestion I made? I prefer to put it explicitly here than to have it implicit in the linked comment.

From the benchmarking we did, TE platform JSON benchmark represents the lowest workload per request. This means the epoll thread will sooner become the bottleneck than on the other benchmarks.

> We should have some heuristics that are good for the general cases

This heuristic is good for the general case, It uses less epoll threads than the previous heuristic (which was a guess with little benchmarking done) and achieves higher performance.

> and allow customer to define custom values that might be better for them.

This is now possible, the count can be set explicitly using the envvar.

Reply:
The env var is good. But the number `28` within the code is my concern. Unless we say we need to pick one, but 28 because CITRINE should not be the reason IMO.

Reply:
@adamsitnik, on the 56-core machine (2-socket 28-core) are the numbers with or without the env vars `COMPlus_Thread_UseAllCpuGroups=1` and `COMPlus_GCCpuGroup=1`? Without those I suspect it would only be using one CPU group and behaving as a single-socket 28-core machine, with those it should try to use both sockets. The numbers seem to be similar to the 28-core machine and it's a bit odd that 2 epoll threads do better there, though there may be other things going on.

Reply:
> but 28 because CITRINE should not be the reason IMO.

We want the value to be high to not have unnecessary epoll threads.
Benchmarks on 28proc x64 (Citrine) showed 1 epoll thread was not a bottleneck.
Benchmarks on 56proc x64 showed 1 epoll thread to be a bottleneck.
So based on these benchmarks: 28 is a safe value, and 56 is not.

Reply:
I agree with @sebastienros that this value is fairly arbitrary, based on the specific hardware we've tested on.  Does the heuristic hold up on machines with a similar number of cores but a different distribution across nodes?  What about when hyperthreading is disabled?  Did we try it with cloud VMs?

If we believe this is strictly better than the heuristic we currently use, ok.  It just feels a lot like "well it works on my machine" :)

Reply:
> on the 56-core machine (2-socket 28-core) are the numbers with or without the env vars

I am always using the default settings (no env vars)

Reply:
It might be common to set those env vars on multi-numa-node machines if the intention is to scale up. Might also be interesting to try the AMD machine with those env vars since it also has multiple numa nodes. Not suggesting for this change or anything but it might provide more insights on heuristics for number of epoll threads.

Reply:
Ahh nevermind, from a brief look it almost looks like all of the the CPU group stuff is disabled on Linux and those env vars may not have any effect.

Reply:
> It might be common to set those env vars on multi-numa-node machines if the intention is to scale up. 

Thanks for pointing this out. I am going to run this config as well.

Reply:
Pls see my latest comment :) there might be more work to do there in the VM, I'm not up-to-date on what's happening there

Reply:
> I agree with @sebastienros that this value is fairly arbitrary, based on the specific (and limited) hardware we've tested on. Does the heuristic hold up on machines with a similar number of cores but a different distribution across nodes? What about when hyperthreading is disabled? Did we try it with cloud VMs?

We see 1 epoll thread is enough to load a 28 core machine (Citrine) with a benchmark that has low threadpool workload vs epoll workload (TE JSON platform).

That's what is captured by `coresPerSingleEpollThread = 28`.

This heuristic also works well in the likely cases that
- ProcessorCount is lower than 28
- for higher workloads per epoll workload

This heuristic isn't tuned for multi-node machines, or machines with 28+++ procs.